Create block: Code quality improvements for the block scaffolding

gziolo opened the pull request:
## Description
Follow-up for #19773.

It addresses comments raised by @aduth during code review. Those are mostly enhancements that improve the code quality.

## Testing details

This package is still not published to npm out of the obvious reasons, so the temporary way to test it looks as follows:

`npx wp-create-block --help`

![Screen Shot 2020-01-21 at 11 31 58](https://user-images.githubusercontent.com/699132/72797597-b4146e00-3c41-11ea-8716-1ab772b4f54b.png)

`npx wp-create-block --version`

![Screen Shot 2020-01-21 at 11 32 47](https://user-images.githubusercontent.com/699132/72797637-c4c4e400-3c41-11ea-980f-a0ce2c5df8bd.png)

`npx wp-create-block` - to see full interactive mode (`npm install` during the process takes some time)

![wp-create-block](https://user-images.githubusercontent.com/699132/72798086-9562a700-3c42-11ea-9021-8278e6a4f46a.gif)

`npx wp-create-block my-quick-block` - to see quick mode 

`npx wp-create-block --template es5` - to scaffold ES5 template

gziolo committed:
Create block: Code quality improvements for the block scaffolding

gziolo commented on packages/create-block/package.json:
```diff
@@ -1,45 +1,45 @@
-  "publishConfig": {
-    "access": "public"
-  }
+	"name": "@wordpress/create-block",
```
@ntwb - something to consider for the linter maybe?

ntwb replied:
There is the [valid-values-name-scope](https://npmpackagejsonlint.org/docs/en/rules/valid-values/valid-values-name-scope) rule to define _allowed scope_, in this case we would want `@wordpress`

ntwb replied:
Ahhh, I see the issue now, indentation, I didn't notice this when looking at just the comment in isolation, this was discussed in https://github.com/tclindner/npm-package-json-lint/issues/81, but at the time there were no great JSON AST parsers, I suspect that has not changed recently

gziolo replied:
Thanks @ntwb for getting back. It seems like copy and paste issue then. I guess it means it was copy and paste issue, I better double check .editorconfig for JSON files in the templates 👍

gziolo committed:
Improve the strucutre and handling of templates
Props to @aduth for the proposal: https://github.com/WordPress/gutenberg/pull/19773#discussion_r370833992.

gziolo committed:
Ensure that package-lock.json file is refreshed with the changes from master

gziolo committed:
Docs: Add a note about version and help options

gziolo committed:
Code style: Run Prettier formatting on the package files

gziolo committed:
Create block: Align .editorconfig with Gutenberg settings

gziolo committed:
Fix: Use the description provided to fill the `description` field in `package.json` file in ESNext template

gziolo committed:
Fix: Ensure that values provided for slug and namespace get converted to lower case

gziolo committed:
Fix: Simplify the logic for transforming slug to title

gziolo commented:
@aduth - just letting you know that I addressed all your feedback. 

I highlighted the most important changes in the CHANGELOG:

### Enhancements

- Removed the code that clears the terminal while the block is scaffolded ([#19867](https://github.com/WordPress/gutenberg/pull/19867)).

### Bug Fixes

- Use the description provided to fill the `description` field in `package.json` file in ESNext template ([#19867](https://github.com/WordPress/gutenberg/pull/19867)).
- Ensure that values provided for slug and namespace get converted to lower case ([#19867](https://github.com/WordPress/gutenberg/pull/19867)).

aduth commented on packages/create-block/lib/index.js:
```diff
@@ -3,72 +3,68 @@
-		'and the name of the WordPress plugin. The rest of the configuration is set to all default values.'
+			'[slug] is optional. When provided it triggers the quick mode where it is used ' +
+			'as the block slug used for its identification, the output location for scaffolded files, ' +
+			'and the name of the WordPress plugin. The rest of the configuration is set to all default values.'
```
I don't necessarily see this as necessary (or correct for that matter)? Seemed fine (correct) before. Not a big deal either way.

gziolo replied:
I executed Prettier on the package out of curiosity, the number of changes applied  was very small 😃this is one of them

aduth commented on packages/create-block/lib/templates.js:
```diff
@@ -24,62 +24,56 @@ const templates = {
-			'$slug.php': 'es5/plugin-php',
-			'style.css': 'style-css',
-		},
+		outputFiles: [ '.editorconfig', 'editor.css', 'index.js', '$slug.php', 'style.css' ],
```
Still think we could just iterate over (recurse into) what exists in the folder itself, rather than explicitly list them (seems redundant). Basically seems like the idea is to copy the template folder, then apply the templating and rename the file to remove the `.mustache` suffix.

Could be done separately.

gziolo replied:
I didn’t fully understood it before. I like the idea a lot and I’ll explore options. Really cool 👍

aduth commented on packages/create-block/lib/templates.js:
```diff
@@ -24,62 +24,56 @@ const templates = {
 		throw new CliError(
-			'Invalid template type name.' +
-			` Allowed values: ${ Object.keys( templates ).join( ', ' ) }.`
+			'Invalid template type name.' + ` Allowed values: ${ Object.keys( templates ).join( ', ' ) }.`
```
Could be simplified?

```suggestion
			`Invalid template type name. Allowed values: ${ Object.keys( templates ).join( ', ' ) }.`
```

gziolo replied:
Sure, Prettier run will create more cases like this.

aduth commented on packages/create-block/CHANGELOG.md:
```diff
@@ -1,5 +1,14 @@
+
+- Use the description provided to fill the `description` field in `package.json` file in ESNext template ([#19867](https://github.com/WordPress/gutenberg/pull/19867)).
+- Ensure that values provided for slug and namespace get converted to lower case ([#19867](https://github.com/WordPress/gutenberg/pull/19867)).
+
```
I guess it can be good to be exhaustive, but technically since we've never published a version before, there's nothing that anyone needs to concern over in the CHANGELOG ?

ntwb replied:
I'm guessing @gziolo will deprecate his package and add a notice linking to this package

gziolo replied:
Indeed, I want to deprecate the old package as soon as this one is published to npm. I wanted to make it clear what has changed between the import and publish. There are two options. I remove those additional  lines or I copy old changelog for full history. What do you think?

ntwb replied:
I'd copy across the [changelog](https://github.com/gziolo/create-wordpress-block/blob/master/CHANGELOG.md) and also add an addition entry highlighting that the package relocated repos for the changelog for this release

gziolo replied:
I followed your suggestion: b6d87122f34e0dccfe19f23fbb6593c8b58a2ce5.

aduth commented on packages/create-block/lib/prompts.js:
```diff
@@ -59,8 +59,7 @@ const dashicon = {
+		return input && input.replace( /dashicon(s)?-/, '' );
 	},
 };
 
```
I missed it the first time around, but just to be aware: There are proposed revisions to the set of block categories being discussed/proposed at #19279.

I have no good solutions to offer, but it would be nice if we could avoid hard-coding this. I guess we can just come back and update it when the time comes.

aduth reviewed (approved):
Nice 👍

gziolo committed:
Update packages/create-block/lib/templates.js

Co-Authored-By: Andrew Duthie <andrew@andrewduthie.com>

gziolo commented:
Thank you for the review! I'll open another PR for the templates. I'm also happy to apply edits to CHANGELO file as soon as I hear your thoughts on the preferred approach.

gziolo commented:
I plan to archive `create-wordpress-block` repository and npm package as soon as this new package is published to npm.
